subsys/mgmt/mcumgr/grp/img_mgmt_client/src/img_mgmt_client.c

```diff
@@ -0,0 +1,617 @@
+	}
+
+	/* Calculate worst case header size for adapt payload length */
+	client->upload.upload_header_size = upload_message_header_size(&client->upload);
+	if (client->upload.upload_header_size) {
+		rc = MGMT_ERR_EOK;
+	} else {
+		rc = MGMT_ERR_ENOMEM;
+	}
+	k_mutex_unlock(&mcumgr_mutex);
+	return rc;
+}
+
+int img_mgmt_client_upload(struct img_mgmt_client *client, const uint8_t *data, size_t length,
+			   struct mcumgr_image_upload *res_buf)
+{
+	struct net_buf *nb;
+	const uint8_t *write_ptr;
+	int rc;
+	uint32_t map_count;
+	bool ok;
+	size_t write_length, max_data_length, offset_before_send, request_length, wrote_length;
+	zcbor_state_t zse[CONFIG_MCUMGR_SMP_CBOR_MAX_DECODING_LEVELS + 2];
+
+	k_mutex_lock(&mcumgr_mutex, K_FOREVER);
+	active_client = client;
+	image_upload_buf = res_buf;
+
+	request_length = length;
+	wrote_length = 0;
+	/* Calculate max data length based on net_buf size - (SMP header + CBOR message_len) */
+	max_data_length = CONFIG_MCUMGR_TRANSPORT_NETBUF_SIZE -
+			  (active_client->upload.upload_header_size + MGMT_HDR_SIZE);
+	/* Trim length based on CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE */
+	if (max_data_length % CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE) {
+		max_data_length -=
+			(max_data_length % CONFIG_MCUMGR_GRP_IMG_UPLOAD_DATA_ALIGNMENT_SIZE);
+	}
+
+	while (request_length != wrote_length) {
+		write_ptr = data + wrote_length;
+		write_length = request_length - wrote_length;
+		if (write_length > max_data_length) {
+			write_length = max_data_length;
+		}
+
+		nb = smp_client_buf_allocation(active_client->smp_client, MGMT_GROUP_ID_IMAGE,
+					       IMG_MGMT_ID_UPLOAD, MGMT_OP_WRITE);
+		if (!nb) {
+			image_upload_buf->status = MGMT_ERR_ENOMEM;
+			goto end;
+		}
+
+		zcbor_new_encode_state(zse, ARRAY_SIZE(zse), nb->data + nb->len,
+				       net_buf_tailroom(nb), 0);
+		if (active_client->upload.offset) {
+			map_count = 6;
+		} else if (active_client->upload.hash_initialized) {
+			map_count = 12;
+		} else {
+			map_count = 10;
+		}
+
+		/* Init map start and write image info and data */
+		ok = zcbor_map_start_encode(zse, map_count) && zcbor_tstr_put_lit(zse, "image") &&
+		     zcbor_uint32_put(zse, active_client->upload.image_num) &&
+		     zcbor_tstr_put_lit(zse, "data") &&
+		     zcbor_bstr_encode_ptr(zse, write_ptr, write_length);
+		/* Write Len when offset is zero */
+		if (ok && !active_client->upload.offset) {
+			ok = zcbor_tstr_put_lit(zse, "len") &&
+			     zcbor_size_put(zse, active_client->upload.image_size);
+		}
+		/* Write offset */
+		if (ok) {
+			ok = zcbor_tstr_put_lit(zse, "off") &&
+			     zcbor_size_put(zse, active_client->upload.offset);
+		}
+		/* Write hash when it defined and offset is 0 */
+		if (ok && !active_client->upload.offset && active_client->upload.hash_initialized) {
+			ok = zcbor_tstr_put_lit(zse, "sha") &&
+			     zcbor_bstr_encode_ptr(zse, active_client->upload.sha256,
+						   IMG_MGMT_HASH_LEN);
```

Comment:
Move inside the above if block and then only need to check if okay and if hash is set instead of 3 conditions